packages/storage-plus/src/snapshot_item.rs

```diff
@@ -0,0 +1,290 @@
+#![cfg(feature = "iterator")]
+
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+
+use cosmwasm_std::{Order, StdError, StdResult, Storage};
+
+use crate::keys::U64Key;
+use crate::snapshot::{ChangeSet, Snapshot};
+use crate::{Bound, Item, Strategy};
+
+/// Item that maintains a snapshot of one or more checkpoints.
+/// We can query historical data as well as current state.
+/// What data is snapshotted depends on the Strategy.
+pub struct SnapshotItem<'a, T> {
+    primary: Item<'a, T>,
+    snapshots: Snapshot<'a, (), T>,
+}
+
+impl<'a, T> SnapshotItem<'a, T> {
+    /// Usage: SnapshotItem::new(snapshot_names!("foobar"), Strategy::EveryBlock)
+    pub const fn new(
+        storage_key: &'a str,
+        checkpoints: &'a str,
+        changelog: &'a str,
+        strategy: Strategy,
+    ) -> Self {
+        SnapshotItem {
+            primary: Item::new(storage_key),
+            snapshots: Snapshot::new(checkpoints, changelog, strategy),
+        }
+    }
+
+    pub fn add_checkpoint(&self, store: &mut dyn Storage, height: u64) -> StdResult<()> {
+        self.snapshots.add_checkpoint(store, height)
+    }
+
+    pub fn remove_checkpoint(&self, store: &mut dyn Storage, height: u64) -> StdResult<()> {
+        self.snapshots.remove_checkpoint(store, height)
+    }
+}
+
+impl<'a, T> SnapshotItem<'a, T>
+where
+    T: Serialize + DeserializeOwned + Clone,
+{
+    /// load old value and store changelog
+    fn write_change(&self, store: &mut dyn Storage, height: u64) -> StdResult<()> {
+        // if there is already data in the changelog for this block, do not write more
+        if self
```

Comment:
likewise, a helper here, like:

`self.snapshots.has_changelog(store, (), height)` would be nice.

Reply:
And maybe expose `write_change` on `self.snapshots`, as this same basic logic is used for Map